### PR TITLE
Closes #8943 refactor SwipeRefreshFeature to not use EngineSession.Observer

### LINKED ISSUE #8943: Refactor SwipeRefreshFeature
We have some issues caused by `SwipeRefreshFeature` using `EngineSession.Observer` 
by default `EngineSession.Observer` rejects [onAppPermissionRequest and onContentPermissionRequest
](https://github.com/mozilla-mobile/android-components/blob/aee6f63bdccd91c486ecd7da9dc4f2601d5cb252/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt#L75), as a side effect of it we have related site permissions issues like https://github.com/mozilla-mobile/fenix/issues/8987 and https://github.com/mozilla-mobile/fenix/issues/16411, moreover eventually we want to remove`EngineSession.Observer`, for the reasons mentioned above is a good a idea to refactor `SwipeRefreshFeature` to not use `EngineSession.Observer` instead use the `BrowserStore`.

### PR DESCRIPTION
**Pull request by Amejia481**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/session/src/main/java/mozilla/components/feature/session/SwipeRefreshFeature.kt [hunk 1]
```diff
@@ -47,21 +42,28 @@ class SwipeRefreshFeature(
...
-                    onLoadingStateChanged(loading)
+                .ifAnyChanged {
+                    arrayOf(it?.content?.loading, it?.content?.refreshCanceled)
+                }
+                .collect { tab ->
+                    tab?.let {
+                        if (!tab.content.loading || tab.content.refreshCanceled) {
+                            swipeRefreshLayout.isRefreshing = false
+                            if (tab.content.refreshCanceled) {
+                                // In case the user tries to refresh again
```
Comment (Mugurell) [inline components/feature/session/src/main/java/mozilla/components/feature/session/SwipeRefreshFeature.kt:55]: Nice 👍

**Review by Mugurell — APPROVED** [review]
LGTM

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/ContentState.kt [hunk 2]
```diff
@@ -46,6 +46,7 @@ import mozilla.components.concept.engine.window.WindowRequest
  * @property loadRequest last [LoadRequestState] if this session.
  * @property permissionRequestsList Holds unprocessed content requests.
  * @property appPermissionRequestsList Holds unprocessed app requests.
+ * @property refreshCanceled Indicates if an intent of refreshing was canceled.
```
Comment (csadilek) [inline components/browser/state/src/main/java/mozilla/components/browser/state/state/ContentState.kt:49]: nit: True if a page refresh was cancelled by the user. Note that this is not about stopping an ongoing page load but useful in cases like swipe-to-refresh which allow users to cancel or abort before a page is refreshed.

Comment (csadilek) [reply components/browser/state/src/main/java/mozilla/components/browser/state/state/ContentState.kt:49]: Let's make this as clear as possible ^^.

**Review by csadilek — APPROVED** [review]
Nice, I think this worked out and cleaned things up at the same time. Just a nit to add a bit more docs.

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/ContentState.kt (hunk 2 above)
Comment (Amejia481) [reply components/browser/state/src/main/java/mozilla/components/browser/state/state/ContentState.kt:49]: 👍🏽

**Pull request merged**